File: Comms Protocol Java/Comms Protocol Java Tests/comms/protocol/java/tests/DataPacketUnitTests.java
package comms.protocol.java.tests;

import org.junit.Test;

import comms.protocol.java.DataPacket;
import comms.protocol.java.Motus_1_RawDataPacket;
import comms.protocol.java.ValidPacketTypes;

import static org.junit.Assert.assertEquals;

public class DataPacketUnitTests 
{
	short testPayloadLen = 0x0005;
    byte[] streamTest = new byte[] { DataPacket.Header1, DataPacket.Header2, (byte)ValidPacketTypes.test_packet.getValue(), (byte)(0x0005 >> 8),
                                (byte)(0x0005), 0, 1, 2, 3, 4 };
	
	@Test
	public void DataPacket_TestConstructor()
	{
		DataPacket defaultConstr = new DataPacket();
        byte[] test = new byte[] { 0, 1, 2, 3, 4 };
        DataPacket constr = new DataPacket(ValidPacketTypes.test_packet, (short)test.length, test);

        assertEquals(ValidPacketTypes.end_valid_packet_types, defaultConstr.getPacketType());
        assertEquals(-1, defaultConstr.getExpectedLen());

        assertEquals(ValidPacketTypes.test_packet, constr.getPacketType());
        assertEquals(test.length, constr.getExpectedLen());
        for (int i = 0; i < constr.getExpectedLen(); i++)
            assertEquals(constr.getPayload()[i], test[i]);
	}
	
	@Test
	public void DataPacket_TestSerializeFromStream()
    {
        DataPacket packet = new DataPacket();
        int index = packet.SerializeFromStream(streamTest, 0);
        assertEquals(streamTest.length, index);
        assertEquals(ValidPacketTypes.test_packet, packet.getPacketType());
        assertEquals(testPayloadLen, packet.getExpectedLen());
        for (int i = 0; i < packet.getExpectedLen(); i++)
        	assertEquals(packet.getPayload()[i], streamTest[DataPacket.DataStartPos + i]);
    }
	
	@Test
	public void DataPacket_TestSerializeFromStreamBadHeader()
    {
        DataPacket packet = new DataPacket();
        int offset = 1;
        offset = packet.SerializeFromStream(streamTest, offset);
        assertEquals(2, offset);
        assertEquals(ValidPacketTypes.end_valid_packet_types, packet.getPacketType());
        assertEquals(-1, packet.getExpectedLen());
    }
	
	@Test
	public void DataPacket_TestSerializeFromStreamBadLen()
    {
        byte[] badLen = new byte[] { DataPacket.Header1, DataPacket.Header2, (byte)ValidPacketTypes.test_packet.getValue(), (byte)(0x0005 >> 8),
                                (byte)(0x0005), 0, 1, 2, 3 };
        DataPacket packet = new DataPacket();
        int offset = 0;
        offset = packet.SerializeFromStream(badLen, offset);
        assertEquals(5, offset);
        assertEquals(ValidPacketTypes.end_valid_packet_types, packet.getPacketType());
        assertEquals(-1, packet.getExpectedLen());
    }
	
	@Test
	public void DataPacket_TestSerializeFromStreamBadType()
    {
        byte[] badType = new byte[] { DataPacket.Header1, DataPacket.Header2, (byte)(ValidPacketTypes.end_valid_packet_types.getValue() + 1), (byte)(0x0005 >> 8),
                                (byte)(0x0005), 0, 1, 2, 3, 4 };
        DataPacket packet = new DataPacket();
        int offset = 0;
        offset = packet.SerializeFromStream(badType, offset);
        assertEquals(3, offset);
        assertEquals(ValidPacketTypes.end_valid_packet_types, packet.getPacketType());
        assertEquals(-1, packet.getExpectedLen());
    }
	
	@Test
	public void DataPacket_TestSerializeToStream()
    {
        byte[] stream = new byte[64];
        byte[] test = new byte[] { 0, 1, 2, 3, 4 };
        DataPacket packet = new DataPacket(ValidPacketTypes.test_packet, (short)test.length, test);
        int numBytesQueued = packet.SerializeToStream(stream, 0);
        assertEquals(streamTest.length, numBytesQueued);
        assertEquals(ValidPacketTypes.test_packet, packet.getPacketType());
        assertEquals(test.length, packet.getExpectedLen());
        for (int i = 0; i < packet.getExpectedLen(); i++)
            assertEquals(packet.getPayload()[i], streamTest[DataPacket.DataStartPos + i]);
    }
	
	@Test
	public void DataPacket_TestSerializeToStreamBadType()
    {
        byte[] stream = new byte[64];
        byte[] test = new byte[] { 0, 1, 2, 3, 4 };
        DataPacket packet = new DataPacket(ValidPacketTypes.end_valid_packet_types, (short)test.length, test);
        int numBytesQueued = packet.SerializeToStream(stream, 0);
        assertEquals(0, numBytesQueued);
        assertEquals(ValidPacketTypes.end_valid_packet_types, packet.getPacketType());
        assertEquals((short)test.length, packet.getExpectedLen());
    }
	
	@Test
	public void DataPacket_TestSerializeToStreamBufferOverRun()
    {
        byte[] stream = new byte[1];
        byte[] test = new byte[] { 0, 1, 2, 3, 4 };
        DataPacket packet = new DataPacket(ValidPacketTypes.test_packet, (short)test.length, test);
        int numBytesQueued = packet.SerializeToStream(stream, 0);
        assertEquals(0, numBytesQueued);
        assertEquals(ValidPacketTypes.test_packet, packet.getPacketType());
        assertEquals((short)test.length, packet.getExpectedLen());
    }
	
}

File: Comms Protocol CSharp/Comms Protocol CSharp Tests/DataPacketUnitTests.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Comms_Protocol_CSharp;

namespace Comms_Protocol_CSharp_Tests
{
    [TestClass]
    public class DataPacketUnitTests
    {
        short testPayloadLen = 0x0005;
        byte[] streamTest = new byte[] { DataPacket.Header1, DataPacket.Header2, (byte)ValidPacketTypes.test_packet, (byte)(0x0005 >> 8),
                                    (byte)(0x0005), 0, 1, 2, 3, 4 };
        [TestMethod]
        public void DataPacket_TestConstructor()
        {
            DataPacket defaultConstr = new DataPacket();
            byte[] test = new byte[] { 0, 1, 2, 3, 4 };
            DataPacket constr = new DataPacket(ValidPacketTypes.test_packet, (short)test.Length, test);

            Assert.AreEqual(ValidPacketTypes.end_valid_packet_types, defaultConstr.Type);
            Assert.AreEqual(-1, defaultConstr.ExpectedLen);

            Assert.AreEqual(ValidPacketTypes.test_packet, constr.Type);
            Assert.AreEqual(test.Length, constr.ExpectedLen);
            for (int i = 0; i < constr.ExpectedLen; i++)
                Assert.AreEqual(constr.Payload[i], test[i]);
        }

        [TestMethod]
        public void DataPacket_TestSerializeFromStream()
        {
            DataPacket packet = new DataPacket();
            int index = packet.SerializeFromStream(streamTest, 0);
            Assert.AreEqual(streamTest.Length, index);
            Assert.AreEqual(ValidPacketTypes.test_packet, packet.Type);
            Assert.AreEqual(testPayloadLen, packet.ExpectedLen);
            for (int i = 0; i < packet.ExpectedLen; i++)
                Assert.AreEqual(packet.Payload[i], streamTest[DataPacket.DataStartPos + i]);
        }

        [TestMethod]
        public void DataPacket_TestSerializeFromStreamBadHeader()
        {
            DataPacket packet = new DataPacket();
            int offset = 1;
            offset = packet.SerializeFromStream(streamTest, offset);
            Assert.AreEqual(2, offset);
            Assert.AreEqual(ValidPacketTypes.end_valid_packet_types, packet.Type);
            Assert.AreEqual(-1, packet.ExpectedLen);
        }

        [TestMethod]
        public void DataPacket_TestSerializeFromStreamBadLen()
        {
            byte[] badLen = new byte[] { DataPacket.Header1, DataPacket.Header2, (byte)ValidPacketTypes.test_packet, (byte)(0x0005 >> 8),
                                    (byte)(0x0005), 0, 1, 2, 3 };
            DataPacket packet = new DataPacket();
            int offset = 0;
            offset = packet.SerializeFromStream(badLen, offset);
            Assert.AreEqual(5, offset);
            Assert.AreEqual(ValidPacketTypes.end_valid_packet_types, packet.Type);
            Assert.AreEqual(-1, packet.ExpectedLen);
        }

        [TestMethod]
        public void DataPacket_TestSerializeFromStreamBadType()
        {
            byte[] badType = new byte[] { DataPacket.Header1, DataPacket.Header2, (byte)(ValidPacketTypes.end_valid_packet_types + 1), (byte)(0x0005 >> 8),
                                    (byte)(0x0005), 0, 1, 2, 3, 4 };
            DataPacket packet = new DataPacket();
            int offset = 0;
            offset = packet.SerializeFromStream(badType, offset);
            Assert.AreEqual(3, offset);
            Assert.AreEqual(ValidPacketTypes.end_valid_packet_types, packet.Type);
            Assert.AreEqual(-1, packet.ExpectedLen);
        }

        [TestMethod]
        public void DataPacket_TestSerializeToStream()
        {
            byte[] stream = new byte[64];
            byte[] test = new byte[] { 0, 1, 2, 3, 4 };
            DataPacket packet = new DataPacket(ValidPacketTypes.test_packet, (short)test.Length, test);
            int numBytesQueued = packet.SerializeToStream(stream, 0);
            Assert.AreEqual(streamTest.Length, numBytesQueued);
            Assert.AreEqual(ValidPacketTypes.test_packet, packet.Type);
            Assert.AreEqual(test.Length, packet.ExpectedLen);
            for (int i = 0; i < packet.ExpectedLen; i++)
                Assert.AreEqual(packet.Payload[i], streamTest[DataPacket.DataStartPos + i]);
        }

        [TestMethod]
        public void DataPacket_TestSerializeToStreamBadType()
        {
            byte[] stream = new byte[64];
            byte[] test = new byte[] { 0, 1, 2, 3, 4 };
            DataPacket packet = new DataPacket(ValidPacketTypes.end_valid_packet_types, (short)test.Length, test);
            int numBytesQueued = packet.SerializeToStream(stream, 0);
            Assert.AreEqual(0, numBytesQueued);
            Assert.AreEqual(ValidPacketTypes.end_valid_packet_types, packet.Type);
            Assert.AreEqual((short)test.Length, packet.ExpectedLen);
        }

        [TestMethod]
        public void DataPacket_TestSerializeToStreamBufferOverRun()
        {
            byte[] stream = new byte[1];
            byte[] test = new byte[] { 0, 1, 2, 3, 4 };
            DataPacket packet = new DataPacket(ValidPacketTypes.test_packet, (short)test.Length, test);
            int numBytesQueued = packet.SerializeToStream(stream, 0);
            Assert.AreEqual(0, numBytesQueued);
            Assert.AreEqual(ValidPacketTypes.test_packet, packet.Type);
            Assert.AreEqual((short)test.Length, packet.ExpectedLen);
        }
    }
}
